```
   Compiling style_tests v0.0.1 (file:///Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style)
error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:209:55
    |
209 |               parsed: ComputedColor::RGBA(RGBA { red: 0f32, green: 0f32, blue: 1f32, alpha: 1f32 }),
    |                                                       ^^^^ expected u8, found f32

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:209:68
    |
209 |               parsed: ComputedColor::RGBA(RGBA { red: 0f32, green: 0f32, blue: 1f32, alpha: 1f32 }),
    |                                                                    ^^^^ expected u8, found f32

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:209:80
    |
209 |               parsed: ComputedColor::RGBA(RGBA { red: 0f32, green: 0f32, blue: 1f32, alpha: 1f32 }),
    |                                                                                ^^^^ expected u8, found f32

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:209:93
    |
209 |               parsed: ComputedColor::RGBA(RGBA { red: 0f32, green: 0f32, blue: 1f32, alpha: 1f32 }),
    |                                                                                             ^^^^ expected u8, found f32

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:213:63
    |
213 |           properties.push(PropertyDeclaration::BorderTopColor(blue.clone()));
    |                                                               ^^^^^^^^^^^^ expected struct `style::values::specified::CSSColor`, found struct `std::boxed::Box`
    |
    = note: expected type `style::properties::DeclaredValue<style::values::specified::CSSColor>`
               found type `style::properties::DeclaredValue<std::boxed::Box<style::values::specified::CSSColor>>`

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:214:65
    |
214 |           properties.push(PropertyDeclaration::BorderRightColor(blue.clone()));
    |                                                                 ^^^^^^^^^^^^ expected struct `style::values::specified::CSSColor`, found struct `std::boxed::Box`
    |
    = note: expected type `style::properties::DeclaredValue<style::values::specified::CSSColor>`
               found type `style::properties::DeclaredValue<std::boxed::Box<style::values::specified::CSSColor>>`

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:215:66
    |
215 |           properties.push(PropertyDeclaration::BorderBottomColor(blue.clone()));
    |                                                                  ^^^^^^^^^^^^ expected struct `style::values::specified::CSSColor`, found struct `std::boxed::Box`
    |
    = note: expected type `style::properties::DeclaredValue<style::values::specified::CSSColor>`
               found type `style::properties::DeclaredValue<std::boxed::Box<style::values::specified::CSSColor>>`

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:216:64
    |
216 |           properties.push(PropertyDeclaration::BorderLeftColor(blue.clone()));
    |                                                                ^^^^^^^^^^^^ expected struct `style::values::specified::CSSColor`, found struct `std::boxed::Box`
    |
    = note: expected type `style::properties::DeclaredValue<style::values::specified::CSSColor>`
               found type `style::properties::DeclaredValue<std::boxed::Box<style::values::specified::CSSColor>>`

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:241:55
    |
241 |               parsed: ComputedColor::RGBA(RGBA { red: 0f32, green: 0f32, blue: 1f32, alpha: 1f32 }),
    |                                                       ^^^^ expected u8, found f32

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:241:68
    |
241 |               parsed: ComputedColor::RGBA(RGBA { red: 0f32, green: 0f32, blue: 1f32, alpha: 1f32 }),
    |                                                                    ^^^^ expected u8, found f32

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:241:80
    |
241 |               parsed: ComputedColor::RGBA(RGBA { red: 0f32, green: 0f32, blue: 1f32, alpha: 1f32 }),
    |                                                                                ^^^^ expected u8, found f32

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:241:93
    |
241 |               parsed: ComputedColor::RGBA(RGBA { red: 0f32, green: 0f32, blue: 1f32, alpha: 1f32 }),
    |                                                                                             ^^^^ expected u8, found f32

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:245:63
    |
245 |           properties.push(PropertyDeclaration::BorderTopColor(blue.clone()));
    |                                                               ^^^^^^^^^^^^ expected struct `style::values::specified::CSSColor`, found struct `std::boxed::Box`
    |
    = note: expected type `style::properties::DeclaredValue<style::values::specified::CSSColor>`
               found type `style::properties::DeclaredValue<std::boxed::Box<style::values::specified::CSSColor>>`

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:246:65
    |
246 |           properties.push(PropertyDeclaration::BorderRightColor(blue.clone()));
    |                                                                 ^^^^^^^^^^^^ expected struct `style::values::specified::CSSColor`, found struct `std::boxed::Box`
    |
    = note: expected type `style::properties::DeclaredValue<style::values::specified::CSSColor>`
               found type `style::properties::DeclaredValue<std::boxed::Box<style::values::specified::CSSColor>>`

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:247:66
    |
247 |           properties.push(PropertyDeclaration::BorderBottomColor(blue.clone()));
    |                                                                  ^^^^^^^^^^^^ expected struct `style::values::specified::CSSColor`, found struct `std::boxed::Box`
    |
    = note: expected type `style::properties::DeclaredValue<style::values::specified::CSSColor>`
               found type `style::properties::DeclaredValue<std::boxed::Box<style::values::specified::CSSColor>>`

error[E0308]: mismatched types
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/properties/serialization.rs:248:64
    |
248 |           properties.push(PropertyDeclaration::BorderLeftColor(blue.clone()));
    |                                                                ^^^^^^^^^^^^ expected struct `style::values::specified::CSSColor`, found struct `std::boxed::Box`
    |
    = note: expected type `style::properties::DeclaredValue<style::values::specified::CSSColor>`
               found type `style::properties::DeclaredValue<std::boxed::Box<style::values::specified::CSSColor>>`
```

--- a/components/style/properties/shorthand/border.mako.rs
+++ b/components/style/properties/shorthand/border.mako.rs
@@ -3,7 +3,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 <%namespace name="helpers" file="/helpers.mako.rs" />
-<% from data import to_rust_ident, ALL_SIDES, maybe_moz_logical_alias %>
+<% from data import to_rust_ident, ALL_SIDES, PHYSICAL_SIDES, maybe_moz_logical_alias %>
 
 ${helpers.four_sides_shorthand("border-color", "border-%s-color", "specified::CSSColor::parse",
                                spec="https://drafts.csswg.org/css-backgrounds/#border-color")}
@@ -141,14 +141,38 @@ pub fn parse_border(context: &ParserContext, input: &mut Parser)
 
     impl<'a> LonghandsToSerialize<'a>  {
         fn to_css_declared<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+            let all_equal = {
+                % for side in PHYSICAL_SIDES:
+                  let border_${side}_width = self.border_${side}_width;
+                  let border_${side}_style = self.border_${side}_style;
+                  let border_${side}_color = self.border_${side}_color;
+                % endfor
+
+                border_top_width == border_right_width &&
+                border_right_width == border_bottom_width &&
+                border_bottom_width == border_left_width &&
+
+                border_top_style == border_right_style &&
+                border_right_style == border_bottom_style &&
+                border_bottom_style == border_left_style &&
+
+                border_top_color == border_right_color &&
+                border_right_color == border_bottom_color &&
+                border_bottom_color == border_left_color
+            };
+
             // If all longhands are all present, then all sides should be the same,
             // so we can just one set of color/style/width
-            super::serialize_directional_border(
-                dest,
-                self.border_${side}_width,
-                self.border_${side}_style,
-                self.border_${side}_color
-            )
+            if all_equal {
+                super::serialize_directional_border(
+                    dest,
+                    self.border_${side}_width,
+                    self.border_${side}_style,
+                    self.border_${side}_color
+                )
+            } else {
+                Ok(())
+            }
         }
     }
 

--- a/tests/unit/style/properties/serialization.rs
+++ b/tests/unit/style/properties/serialization.rs
@@ -187,6 +187,72 @@ mod shorthand_serialization {
         }
 
         #[test]
+        fn different_longhands_should_serialize_to_long_form() {
+          let mut properties = Vec::new();
+
+          let solid = DeclaredValue::Value(BorderStyle::solid);
+
+          properties.push(PropertyDeclaration::BorderTopStyle(solid.clone()));
+          properties.push(PropertyDeclaration::BorderRightStyle(solid.clone()));
+          properties.push(PropertyDeclaration::BorderBottomStyle(solid.clone()));
+          properties.push(PropertyDeclaration::BorderLeftStyle(solid.clone()));
+
+          let px_30 = DeclaredValue::Value(BorderWidth::from_length(Length::from_px(30f32)));
+          let px_10 = DeclaredValue::Value(BorderWidth::from_length(Length::from_px(10f32)));
+
+          properties.push(PropertyDeclaration::BorderTopWidth(px_30.clone()));
+          properties.push(PropertyDeclaration::BorderRightWidth(px_30.clone()));
+          properties.push(PropertyDeclaration::BorderBottomWidth(px_30.clone()));
+          properties.push(PropertyDeclaration::BorderLeftWidth(px_10.clone()));
+
+          let blue = DeclaredValue::Value(CSSColor {
+              parsed: ComputedColor::RGBA(RGBA::new(0, 0, 255, 255)),
+              authored: None
+          });
+
+          properties.push(PropertyDeclaration::BorderTopColor(blue.clone()));
+          properties.push(PropertyDeclaration::BorderRightColor(blue.clone()));
+          properties.push(PropertyDeclaration::BorderBottomColor(blue.clone()));
+          properties.push(PropertyDeclaration::BorderLeftColor(blue.clone()));
+
+          let serialization = shorthand_properties_to_string(properties);
+          assert_eq!(serialization,
+          "border-style: solid; border-width: 30px 30px 30px 10px; border-color: rgb(0, 0, 255);");
+        }
+
+        #[test]
+        fn same_longhands_should_serialize_correctly() {
+          let mut properties = Vec::new();
+
+          let solid = DeclaredValue::Value(BorderStyle::solid);
+
+          properties.push(PropertyDeclaration::BorderTopStyle(solid.clone()));
+          properties.push(PropertyDeclaration::BorderRightStyle(solid.clone()));
+          properties.push(PropertyDeclaration::BorderBottomStyle(solid.clone()));
+          properties.push(PropertyDeclaration::BorderLeftStyle(solid.clone()));
+
+          let px_30 = DeclaredValue::Value(BorderWidth::from_length(Length::from_px(30f32)));
+
+          properties.push(PropertyDeclaration::BorderTopWidth(px_30.clone()));
+          properties.push(PropertyDeclaration::BorderRightWidth(px_30.clone()));
+          properties.push(PropertyDeclaration::BorderBottomWidth(px_30.clone()));
+          properties.push(PropertyDeclaration::BorderLeftWidth(px_30.clone()));
+
+          let blue = DeclaredValue::Value(CSSColor {
+              parsed: ComputedColor::RGBA(RGBA::new(0, 0, 255, 255)),
+              authored: None
+          });
+
+          properties.push(PropertyDeclaration::BorderTopColor(blue.clone()));
+          properties.push(PropertyDeclaration::BorderRightColor(blue.clone()));
+          properties.push(PropertyDeclaration::BorderBottomColor(blue.clone()));
+          properties.push(PropertyDeclaration::BorderLeftColor(blue.clone()));
+
+          let serialization = shorthand_properties_to_string(properties);
+          assert_eq!(serialization, "border: 30px solid rgb(0, 0, 255);");
+        }
+
+        #[test]
         fn padding_should_serialize_correctly() {
             let mut properties = Vec::new();
 

